lib/block/merkleblock.js

```diff
@@ -0,0 +1,280 @@
+ */
+MerkleBlock.prototype.toObject = function toObject() {
+  return {
+    header: this.header.toObject(),
+    numTransactions: this.numTransactions,
+    hashes: this.hashes,
+    flags: this.flags
+  };
+};
+
+/**
+ * @returns {String} - A JSON string of a MerkleBlock
+ */
+MerkleBlock.prototype.toJSON = function toJSON() {
+  return JSON.stringify(this.toObject());
+};
+
+/**
+ * Verify that the MerkleBlock is valid
+ * @returns {Bool} - True/False whether this MerkleBlock is Valid
+ */
+MerkleBlock.prototype.validMerkleTree = function validMerkleTree() {
+  $.checkState(this.flags instanceof Array, 'MerkleBlock flags is not an array');
+  $.checkState(this.hashes instanceof Array, 'MerkleBlock flags is not an array');
+  var self = this;
+  if(this._validMerkleTree === true) {
+    return true;
+  } else if (this._validMerkleTree === false) {
+    return false;
+  }
+
+  // Can't have more hashes than numTransactions
+  if(this.hashes.length > this.numTransactions) {
+    return this._setValidMerkleTree(false);
+  }
+
+  // Can't have more flag bits than num hashes
+  if(this.flags.length * 8 < this.hashes.length) {
+    return this._setValidMerkleTree(false);
+  }
+
+  // Calculate height of tree
+  // From Bitcoin Core merkleblock.h CalcTreeWidth() + CPartialMerkleTree
+  var height = 0;
+  while (calcTreeWidth(height) > 1) {
+    height++;
+  }
+
+  var flagBitsUsed = 0;
+  var hashesUsed = 0;
+
+  // Modeled after Bitcoin Core merkleblock.h CalcTreeWidth()
+  function calcTreeWidth(height) {
+    return (self.numTransactions + (1 << height) - 1) >> height;
+  }
+
+  // Modeled after Bitcoin Core merkleblock.cpp TraverseAndExtract()
+  function traverse(depth, pos) {
+    if(flagBitsUsed > self.flags.length * 8) {
+      return null;
+    }
+    var isParentOfMatch = (self.flags[flagBitsUsed >> 3] >>> (flagBitsUsed++ & 7)) & 1;
+    if(depth === 0 || !isParentOfMatch) {
+      if(hashesUsed >= self.hashes.length) {
+        return null;
+      }
+      var hash = self.hashes[hashesUsed++];
+      return new Buffer(hash, 'hex');
+    } else {
+      var left = traverse(depth-1, pos*2);
+      var right;
+      if(pos*2+1 < calcTreeWidth(depth-1)) {
+        right = traverse(depth-1, pos*2+1);
+      } else {
+        right = left;
+      }
+      return Hash.sha256sha256(new Buffer.concat([left, right]));
+    }
+  }
+
+  var root = traverse(height, 0);
+  if(hashesUsed !== this.hashes.length) {
+    return this._setValidMerkleTree(false);
+  }
+  return this._setValidMerkleTree(BufferUtil.equals(root, this.header.merkleRoot));
+}
+
+/**
+ * @param {Transaction|String} - Transaction or Transaction ID Hash
+ * @returns {Bool} - return true/false if this MerkleBlock has the TX or not
+ * @private
+ */
+MerkleBlock.prototype.hasTransaction = function hasTransaction(tx) {
+  $.checkArgument(!_.isUndefined(tx), 'No transaction given');
+  $.checkArgument(tx instanceof Transaction
+    || typeof tx === 'string', 'No transaction given');
```

Comment:
Please use different error messages for different errors. This allows for better testing. I'd use `'Wrong transaction type given, use "Transaction" or "string"`